### PR TITLE
Lin/flpy 6 version check

### PR DESCRIPTION
**Pull request by jinlin-flex**
[Jira Ticket](https://flow360.atlassian.net/browse/FLPY-6?atlOrigin=eyJpIjoiOTc5YjAxOTFlNzg4NGM2ZDhlYzYzYTM4MzI4N2Q3YzgiLCJwIjoiaiJ9)

Summary:
Implement version check to get information on if the client version is
1)supported
2)the latest version
3)throw SystemExit 0 error if client version is not supported

Coverage:
flow360/solver_version.py 27 5 81%
flow360/version_check.py 51 0 100%

Note:
python client version follows [PEP440](https://peps.python.org/pep-0440/) semantics
other app versions follow solver_version.py semantics

Future TODO:
Use standard python version format [PEP440](https://peps.python.org/pep-0440/) semantics for all version naming

### CODE REVIEW

File: flow360/version_check.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    """
+
+    YES = 1
+    NO = 2
+    CAN_UPGRADE = 3
+
+
+def get_supported_server_versions(app_name):
+    """
+    fetch list of supported versions
```
Comment (lei-flex) [inline flow360/version_check.py:27]: same above.:)

File: flow360/version_check.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    try:
+        session = requests.Session()
+        http = http_util.Http(session)
+        response = http.get(f"versions?app_name={app_name}")
+    except HTTPError as exc:
```
Comment (lei-flex) [inline flow360/version_check.py:33]: except HTTPError as exc -->  except HTTPError as error:

exc is not good abbreviation.

File: flow360/version_check.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        raise HTTPError("Error in connecting server") from exc
+
+    versions = [re.sub(r".+-", "", item["version"]) for item in response]
+
+    if len(versions) == 0:
+        raise RuntimeError("Error in fetching supported versions")
```
Comment (lei-flex) [inline flow360/version_check.py:39]: the message should be " the supported versions should not be empty for {app_name}"

File: flow360/version_check.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    return versions
+
+
+def check_client_version(app_name):
```
Comment (lei-flex) [inline flow360/version_check.py:44]: not good python3 function defination:
for example:   def save(slc: Slice, commit: bool = True) -> None:  as good reference.

Comment (jinlin-flex) [reply flow360/version_check.py:44]: Should the tests also have python3 function definition?

File: flow360/version_check.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    if not is_supported:
+        return VersionSupported.NO, current_version
+
+    if can_upgrade:
+        return VersionSupported.CAN_UPGRADE, latest_version
+
+    return VersionSupported.YES, current_version
+
+
+def client_version_get_info(app_name: str) -> None:
```
Comment (feilin-flexcompute) [inline flow360/version_check.py:83]: Could you point me to the place where this function is called? I can not find it.

File: flow360/version_check.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
+"""Client version check module"""
```
Comment (lei-flex) [inline flow360/version_check.py:1]: """
This module provides utility functions for mathematical operations.

Functions:
- add(x, y): Returns the sum of two numbers.
- subtract(x, y): Returns the difference between two numbers.
- multiply(x, y): Returns the product of two numbers.
- divide(x, y): Returns the division of two numbers.

Constants:
- PI: The mathematical constant pi.
- EULER: The mathematical constant e.
"""
use it as good example for description.

File: flow360/version_check.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+
+class VersionSupported(Enum):
+    """
+    Enumeration representing the different states of version support.
+    """
+
+    YES = 1
+    NO = 2
+    CAN_UPGRADE = 3
```
Comment (maciej-flexcompute) [inline flow360/version_check.py:40]: can we use enums:
YES = 'yes', etc?

File: flow360/version_check.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        session = requests.Session()
+        http = http_util.Http(session)
+        response = http.get(f"versions?app_name={app_name}")
+    except HTTPError as error:
+        raise HTTPError(f"failed to retrieve the versions for {app_name}") from error
+
+    versions = [re.sub(r".+-", "", item["version"]) for item in response]
+
+    if len(versions) == 0:
+        raise RuntimeError(f"the supported versions should not be empty for {app_name}")
```
Comment (maciej-flexcompute) [inline flow360/version_check.py:63]: Lets rewrite this massage:
f"Something went wrong when checking python client version. The supported versions should not be empty for {app_name}. If you see this message again, contact support."

File: flow360/version_check.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+    return versions
+
+
+def check_client_version(app_name: str) -> Tuple[VersionSupported, Union[str, Version]]:
```
Comment (maciej-flexcompute) [inline flow360/version_check.py:68]: app_name is always "flow360-python-client-v2". We don't need to support other app_names. Just use it as const or directly in webAPI call.

File: flow360/version_check.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        app_name (str): The name of the application
+
+    Returns:
+        Tuple[VersionSupported, Union[str, Version]]: A tuple containing the version status
+        and the current or latest version
+    """
+    supported_versions = get_supported_server_versions(app_name)
+    latest_version = supported_versions[0]
+    current_version = __version__
+    if app_name == "flow360-python-client-v2":
```
Comment (maciej-flexcompute) [inline flow360/version_check.py:82]: app_name will be always flow360-python-client-v2. No need for else statement.

File: flow360/version_check.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+    msg = """
+    To upgrade run:
+        pip3 install -U flow360client
+
+    """
+    log.info(msg)
+
+    if version_status == VersionSupported.NO:
+        log.error("Flow360 version not supported")
```
Comment (maciej-flexcompute) [inline flow360/version_check.py:128]: log.error("This flow360 version is no longer supported. Please upgrade or contact support.")

File: tests/test_version_check.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+import flow360.version_check as vc
+
+
+def test_get_supported_server_versions():
+    # Prepare mock data
+    mock_response = MagicMock()
+    mock_response.__iter__.return_value = iter([{"version": "1.0.0"}, {"version": "2.0.3b5"}])
+    # Mock the http_util.Http.get method
+    with patch("flow360.version_check.http_util.Http.get") as mock_get:
```
Comment (maciej-flexcompute) [inline tests/test_version_check.py:14]: I like this mock webapi testing very much! Can you integrate its code with https://github.com/flexcompute/Flow360/blob/develop/tests/mock_server.py

File: flow360/version_check.py (hunk 9 above)
Comment (maciej-flexcompute) [reply flow360/version_check.py:68]: Run `check_client_version()` on import. Add option to skip check - maybe by ENV variable now

**Comment by jinlin-flex** [comment]
I have added the feature to Run check_client_version() upon import. However, currently the API key is not configured in BasicUserConfig(), so  trying to get the versions will raise an AuthorizationError. I have set the FLOW360_SKIP_VERSION_CHECK = False for now, but will this impact future usage of check_client_version() ?

File: flow360/cloud/http_util.py [hunk 13]
```diff
@@ -76,6 +76,18 @@ class Http:
     def __init__(self, session: requests.Session):
         self.session = session
 
+    @http_interceptor
+    def get_portal_api(self, path: str, json=None, params=None):
```
Comment (maciej-flexcompute) [inline flow360/cloud/http_util.py:80]: get_portal_api -> portal_api_get

File: flow360/environment.py [hunk 14]
```diff
@@ -55,6 +55,15 @@ def get_real_url(self, path: str):
...
 )
 
+portal_api_dev = EnvironmentConfig(
+    name="portal_api_dev",
+    web_api_endpoint="https://portal-api.dev-simulation.cloud",
+    aws_region="us-east-1",
+    apikey_profile="dev",
+)
+
+FLOW360_SKIP_VERSION_CHECK = False
```
Comment (maciej-flexcompute) [inline flow360/environment.py:65]: add parameter to `EnvironmentConfig`,
`portal_web_api_endpoint = "https://..."`
Dev:
`portal_web_api_endpoint="https://portal-api.dev-simulation.cloud"`,
Uat:
`portal_web_api_endpoint="https://uat-portal-api.simulation.cloud"`,
Prod:
`portal_web_api_endpoint="https://portal-api.dev-simulation.cloud"`,

Use it:
```
url=Env.dev.get_portal_real_url(path), json=json, params=params, auth=api_key_auth
```

File: flow360/version_check.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+    msg = """
+    To upgrade run:
+        pip3 install -U flow360client
+
+    """
+    log.info(msg)
+
+
+if FLOW360_SKIP_VERSION_CHECK:
```
Comment (maciej-flexcompute) [inline flow360/version_check.py:122]: if not FLOW360_SKIP_VERSION_CHECK

File: tests/data/mock_webapi/versions.json [hunk 16]
```diff
@@ -0,0 +1,3 @@
+{
+    "data": []
```
Comment (maciej-flexcompute) [inline tests/data/mock_webapi/versions.json:2]: why are we returning empty response in test?

File: tests/mock_server.py [hunk 17]
```diff
@@ -23,6 +23,17 @@ def raise_for_status(self):
...
 
 
+class MockResponseVersions(MockResponse):
+    """response if get_supported_server_versions()"""
+
+    @staticmethod
+    def json():
+        with open(os.path.join(here, "data/mock_webapi/versions.json")) as fh:
+            res = json.load(fh)
+            res["data"] = [item for item in res["data"]]
```
Comment (maciej-flexcompute) [inline tests/mock_server.py:33]: what is this line doing?

File: flow360/environment.py [hunk 18]
```diff
@@ -103,6 +124,14 @@ def prod(self):
...
         return prod
 
+    @property
+    def portal_api_dev(self):
+        """
+        Get the dev portal api environment.
+        :return:
+        """
+        return portal_api_dev
+
```
Comment (maciej-flexcompute) [inline flow360/environment.py:134]: I think this is not used, remove

File: flow360/environment.py [hunk 19]
```diff
@@ -33,28 +34,48 @@ def get_real_url(self, path: str):
...
     aws_region="us-gov-west-1",
     apikey_profile="default",
 )
 
+portal_api_dev = EnvironmentConfig(
+    name="portal_api_dev",
+    web_api_endpoint="https://portal-api.dev-simulation.cloud",
+    aws_region="us-east-1",
+    apikey_profile="dev",
+)
```
Comment (maciej-flexcompute) [inline flow360/environment.py:75]: I think this is not used, remove

**Review by maciej-flexcompute — APPROVED** [review]

**Pull request merged**